### PR TITLE
Allow archive artifacts to accept lazy task providers

### PR DESCRIPTION
**Pull request by ghale**
This introduces an `ArchiveProviderPublishArtifact` which can accept a task provider instead of a concrete task when configuring artifacts.

This also converts several artifact types to be created lazily and use this artifact type.

CI: https://builds.gradle.org/project.html?projectId=Gradle&tab=projectOverview&branch_Gradle=td%2Ftask-creation-buildops-flag

### CODE REVIEW

**Comment by adammurdoch** [comment]
I'd think about using a different approach. Instead of adding an artifact implementation that can deal with a lazy task and creating such artifacts eagerly, I would reuse the lazy registration API to _lazily_ register an artifact and reuse the existing artifact implementation.

Either way, we'll also need a public API to allow plugins to define outgoing artifacts using the outputs of lazy tasks.

**Comment by adammurdoch** [comment]
Also, we should make sure the java library plugin uses the same approach to publishing the output of a task.

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/publish/ArchivePublishArtifact.java [hunk 1]
```diff
@@ -13,95 +13,21 @@
...
-    private Date date;
-    private File file;
-
-    private AbstractArchiveTask archiveTask;
+public class ArchivePublishArtifact extends AbstractArchivePublishArtifact {
+    private final AbstractArchiveTask archiveTask;
 
     public ArchivePublishArtifact(AbstractArchiveTask archiveTask) {
         super(archiveTask);
         this.archiveTask = archiveTask;
```
Comment (lacasseio) [inline subprojects/core/src/main/java/org/gradle/api/internal/artifacts/publish/ArchivePublishArtifact.java]: As Adam mentioned, I think we could merge this class with the `ArchiveProviderPublishArtifact`. We can converge to the common denominator with `Providers.of(...)`.

**Comment by ghale** [comment]
I've changed this so that it reuses `ArchivePublishArtifact` via `LazyPublishArtifact`.  For now, this just works with a provider of `AbstractArchiveTask`.  I created https://github.com/gradle/gradle-native/issues/723 to handle any arbitrary task using its outputs.

File: subprojects/build-comparison/src/test/groovy/org/gradle/api/plugins/buildcomparison/outcome/internal/tooling/PublishArtifactToFileBuildOutcomeTransformerTest.groovy [hunk 2]
```diff
@@ -63,16 +65,44 @@ class PublishArtifactToFileBuildOutcomeTransformerTest extends AbstractProjectBu
...
+    }
+
     def "can handle generic publish artifact"() {
         given:
         def task = Mock(Task)
         def taskDependency = Mock(TaskDependency)
         def artifact = Mock(PublishArtifact)
 
-        1 * taskDependency.getDependencies(null) >>> [[task] as Set]
+        2 * taskDependency.getDependencies(null) >>> [[task] as Set]
```
[the remark this answers is not in the document]
Comment (ghale) [reply subprojects/build-comparison/src/test/groovy/org/gradle/api/plugins/buildcomparison/outcome/internal/tooling/PublishArtifactToFileBuildOutcomeTransformerTest.groovy]: It's _probably_ not a big deal, but there's no reason for it - we can easily only check it once.  Fixed.

**Pull request merged**